gtsam/slam/tests/testEssentialMatrixFactor.cpp

```diff
@@ -351,7 +353,117 @@ TEST (EssentialMatrixFactor3, minimization) {
   EXPECT_DOUBLES_EQUAL(0, graph.error(result), 1e-4);
 }
 
-} // namespace example1
+//*************************************************************************
+TEST(EssentialMatrixFactor4, factor) {
+  Key keyE(1);
+  Key keyK(1);
+  for (size_t i = 0; i < 5; i++) {
+    EssentialMatrixFactor4<Cal3_S2> factor(keyE, keyK, pA(i), pB(i), model1);
+
+    // Check evaluation
+    Vector1 expected;
+    expected << 0;
+    Matrix HEactual;
+    Matrix HKactual;
+    Vector actual = factor.evaluateError(trueE, trueK, HEactual, HKactual);
+    EXPECT(assert_equal(expected, actual, 1e-7));
+
+    // Use numerical derivatives to calculate the expected Jacobian
+    Matrix HEexpected;
+    Matrix HKexpected;
+    typedef Eigen::Matrix<double, 1, 1> Vector1;
+    boost::function<Vector(const EssentialMatrix &, const Cal3_S2 &)> f =
+        boost::bind(&EssentialMatrixFactor4<Cal3_S2>::evaluateError, factor, _1,
+                    _2, boost::none, boost::none);
+    HEexpected = numericalDerivative21<Vector1, EssentialMatrix, Cal3_S2>(
+        f, trueE, trueK);
+    HKexpected = numericalDerivative22<Vector1, EssentialMatrix, Cal3_S2>(
+        f, trueE, trueK);
+
+    // Verify the Jacobian is correct
+    EXPECT(assert_equal(HEexpected, HEactual, 1e-8));
+    EXPECT(assert_equal(HKexpected, HKactual, 1e-8));
+  }
+}
+
+//*************************************************************************
+TEST(EssentialMatrixFactor4, evaluateErrorJacobians) {
+  Key keyE(1);
+  Key keyK(2);
+  // initialize essential matrix
+  Rot3 r = Rot3::Expmap(Vector3(M_PI / 6, M_PI / 3, M_PI / 9));
+  Unit3 t(Point3(2, -1, 0.5));
+  EssentialMatrix E = EssentialMatrix::FromRotationAndDirection(r, t);
+  Cal3_S2 K(200, 1, 1, 10, 10);
+  Values val;
+  val.insert(keyE, E);
+  val.insert(keyK, K);
+
+  Point2 pA(10.0, 20.0);
+  Point2 pB(12.0, 15.0);
+
+  EssentialMatrixFactor4<Cal3_S2> f(keyE, keyK, pA, pB, model1);
+  EXPECT_CORRECT_FACTOR_JACOBIANS(f, val, 1e-5, 1e-6);
+}
+
+//*************************************************************************
+TEST(EssentialMatrixFactor4, minimization) {
+  // As before, we start with a factor graph and add constraints to it
+  NonlinearFactorGraph graph;
+  for (size_t i = 0; i < 5; i++)
+    graph.emplace_shared<EssentialMatrixFactor4<Cal3_S2>>(1, 2, pA(i), pB(i),
+                                                          model1);
+
+  // Check error at ground truth
+  Values truth;
+  truth.insert(1, trueE);
+  truth.insert(2, trueK);
+  EXPECT_DOUBLES_EQUAL(0, graph.error(truth), 1e-8);
+
+  // Check error at initial estimate
+  Values initial;
+  EssentialMatrix initialE =
+      trueE.retract((Vector(5) << 0.1, -0.1, 0.1, 0.1, -0.1).finished());
+  Cal3_S2 initialK =
+      trueK.retract((Vector(5) << 0.1, -0.08, 0.01, -0.05, 0.06).finished());
```

Comment:
I tried different things here. 
If you add the prior on trueK instead of initialK and start from trueK, the test passes (note initialE != trueE).  
But if the prior is on initialK, K just does not change (even if the covariance is sigma on prior is super high). And E does not converge to the right solution either (which makes sense, as K is incorrect). 

With 5 points, we can only hope to optimize E right? Maybe we should add another test, where we have 8 points and the prior on initialK (not trueK).

Reply:
Yeah, actually, this makes sense: the prior disambiguates in favor of the K that's given. If we are only given 5 points, there is no redundancy and all 5 DOF are devoted to (R, d), also 5DOF.

I agree that with a new test we can check that in the case of > 5 correspondences the calibration is nudged toward the correct K. But, we will never get more than 7DOF from two-view correspondences, so the best we can hope for is that 2DOF of the calibration (any 2D linear combination) is corrected. Typically, we want the focal length to be corrected most, and hence constrain that the least, I'd say.